File: python/linked-list/linked_list.py
class Node(object):
    def __init__(self, value, next=None, previous=None):
        self.value = value
        self.next = next
        self.previous = previous


class LinkedList(object):
    def __init__(self):
        self.head = self.tail = None

    def __len__(self):
        length = 0
        node = self.head
        while node:
            length += 1
            node = node.next
        return length

    def __iter__(self):
        values = []
        node = self.head
        while node:
            values.append(node.value)
            node = node.next
        return iter(values)

    def push(self, value):
        node = Node(value)

        if self.tail:
            self.tail.next = node
            node.previous = self.tail
            self.tail = node
        else:
            self.head = self.tail = node

    def pop(self):
        result = self.tail.value
        new_tail = self.tail.previous
        if new_tail:
            new_tail.next = None
            self.tail = new_tail
        else:
            self.head = self.tail = None
        return result

    def shift(self):
        result = self.head.value
        new_head = self.head.next
        if new_head:
            new_head.previous = None
            self.head = new_head
        else:
            self.head = self.tail = None
        return result

    def unshift(self, value):
        node = Node(value)

        if self.head:
            node.next = self.head
            self.head.previous = node
            self.head = node
        else:
            self.head = self.tail = node

File: javascript/linked-list/linked-list.js
//
// This is only a SKELETON file for the 'Linked List' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class LinkedList {
  constructor() {
    this.head = this.tail = null;
  }

  push(value) {
    const node = new Node(value);

    if (this.tail) {
      this.tail.next = node;
      node.prev = this.tail;
      this.tail = node;
    } else {
      this.head = this.tail = node;
    }
  }

  pop() {
    const result = this.tail.value;

    const newTail = this.tail.prev;
    if (newTail) {
      newTail.next = null;
      this.tail = newTail;
    } else {
      this.head = this.tail = null;
    }

    return result;
  }

  shift() {
    const result = this.head.value;

    const newHead = this.head.next;
    if (newHead) {
      newHead.prev = null;
      this.head = newHead;
    } else {
      this.head = this.tail = null;
    }

    return result;
  }

  unshift(value) {
    const node = new Node(value);

    if (this.head) {
      node.next = this.head;
      this.head.prev = node;
      this.head = node;
    } else {
      this.head = this.tail = node;
    }
  }

  delete(value) {
    let node = this.head;
    while (node && node.value != value) {
      node = node.next;
    }
    if (!node) {
      return;
    }

    if (node == this.head) {
      this.shift();
    } else if (node == this.tail) {
      this.pop();
    } else {
      const prevNode = node.prev;
      const nextNode = node.next;
      prevNode.next = nextNode;
      nextNode.prev = prevNode;
    }
  }

  count() {
    let length = 0;
    let node = this.head;
    while (node) {
      length++;
      node = node.next;
    }

    return length;
  }
}

class Node {
  constructor(value, next, prev) {
    this.value = value;
    this.next = next;
    this.prev = prev;
  }
}
